mesh/node.c

```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+				return false;
+
+			id |= VENDOR_ID_MASK;
+			break;
+		}
+
+		case 8: {
+			if (sscanf(jidx, "%08x", &id) != 1)
+				return false;
+
+			mod->vendor = true;
+			break;
+		}
+
+		default:
+			return false;
+		}
+
+		mod->id = id;
+		l_queue_push_tail(ele->models, mod);
+		json_object_iter_next(&iter);
+	}
+
+	return true;
+}
+
+static bool parse_imported_elements(json_object *jelements,
+						struct mesh_db_node *db_node)
+{
+	json_object *jvalue;
+	json_object *jtemp;
+	struct json_object_iterator iter, end;
+	struct mesh_db_element *ele;
+	const char *jidx;
+
+	if (!json_object_object_get_ex(jelements, "elements", &jvalue))
+		return false;
+
+	if (json_object_get_type(jelements) != json_type_object)
+		return false;
+
+	if (jvalue && json_object_get_type(jvalue) != json_type_object)
+		return false;
+
+	db_node->elements = l_queue_new();
+
+	if (!db_node->elements)
+		return false;
+
+	iter = json_object_iter_begin(jvalue);
+	end = json_object_iter_end(jvalue);
+
+	while (!json_object_iter_equal(&iter, &end)) {
+
+		json_object *jmodels, *jlocation;
+
+		char *str_end, *str;
+		uint8_t idx;
 
+		jidx = json_object_iter_peek_name(&iter);
+		jtemp = json_object_iter_peek_value(&iter);
+
+		if (!*jidx)
+			return false;
+
+		idx = (uint8_t)strtol(jidx, &str_end, 10);
+
+		if (*str_end)
+			return false;
+
+		if (json_object_get_type(jtemp) != json_type_object)
+			return false;
+
+		ele = l_new(struct mesh_db_element, 1);
+
+		if (!ele)
+			return false;
+
+		ele->index = idx;
+		ele->models = l_queue_new();
+
+		if (!ele->models)
+			return false;
+
+		if (!json_object_object_get_ex(jtemp, "location", &jlocation))
+			return false;
+
+		str = (char *)json_object_get_string(jlocation);
+
+		if (sscanf(str, "%04hx", &(ele->location)) != 1)
+			return false;
+
+		if (json_object_object_get_ex(jtemp, "models", &jmodels)) {
+
+			if (json_object_get_type(jmodels) != json_type_object)
+				return false;
+
+			if (!parse_imported_models(jmodels, ele))
+				return false;
+		}
+
+		l_queue_push_tail(db_node->elements, ele);
+		json_object_iter_next(&iter);
+	}
+
+	return true;
+}
+
+static void put_net_keyrings(void *net_keys, void *_node)
+{
+	struct mesh_node *node = _node;
+	struct keyring_net_key *net_key = net_keys;
+
+	if (!node || !net_key)
+		return;
+
+	if (!keyring_put_net_key(node, net_key->net_idx, net_key))
+		return;
+}
+
+static bool parse_imported_finalize(struct mesh_db_node *db_node,
+							struct mesh_node *node)
+{
+	struct mesh_net *net;
+
+	if (!node_init_from_storage(node, db_node)) {
+		node_remove(node);
+		return false;
+	}
+
+	net = node_get_net(node);
+
+	if (!IS_UNASSIGNED(db_node->unicast) &&
+			!mesh_net_register_unicast(net, db_node->unicast,
+			node_get_num_elements(node))) {
+
+		node_remove(node);
+		return false;
+	}
+
+	if (node_uuid_get(node))
+		mesh_net_id_uuid_set(net, node_uuid_get(node));
+
+	return true;
+}
+
+static void add_keys(void *net_keys, void *_node)
+{
+	l_info("ADD KEYS CB");
```

Comment:
`mesh/node.c:1725:2: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

Reply:
log removed